harness/determined/_model_metric_manager.py

```diff
@@ -0,0 +1,231 @@
+            "end_batch": self.end_batch,
+            "start_time": self.start_time.astimezone(timezone.utc).strftime(
+                "%Y-%m-%dT%H:%M:%S.%fZ"
+            ),
+            "state": self.state,
+        }
+
+        r = req.post(
+            self.master_url,
+            "/api/v1/training_metrics",
+            body={"training_metrics": training_metrics},
+        )
+        if not hasattr(r, "headers"):
+            raise Exception(r)
+        return self
+
+    def complete(self, metrics: Dict[str, Any]) -> None:
+        logging.info(
+            f"Finishing saving training metrics for trial {self.trial_id} "
+            f"step {self.step_id} batch {self.start_batch} to {self.end_batch}."
+        )
+        start_time = self.start_time.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        end_time = (
+            datetime.now(timezone.utc).astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        )
+        training_metrics = {
+            "experiment_id": self.experiment_id,
+            "trial_id": self.trial_id,
+            "step_id": self.step_id,
+            "start_batch": self.start_batch,
+            "end_batch": self.end_batch,
+            "start_time": start_time,
+            "end_time": end_time,
+            "state": "STATE_COMPLETED",
+            "metrics": util.make_serializable(metrics),
+        }
+
+        r = req.put(
+            self.master_url,
+            "/api/v1/training_metrics",
+            body={"training_metrics": training_metrics},
+        )
+        if not hasattr(r, "headers"):
+            raise Exception(r)
+
+        self.state = "STATE_COMPLETED"
+
+    def failed(self) -> None:
+        logging.info(
+            f"Failed saving training metrics for trial {self.trial_id} "
+            f"step {self.step_id} batch {self.start_batch} to {self.end_batch}."
+        )
+        start_time = self.start_time.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        end_time = (
+            datetime.now(timezone.utc).astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+        )
+        training_metrics = {
+            "experiment_id": self.experiment_id,
+            "trial_id": self.trial_id,
+            "step_id": self.step_id,
+            "start_batch": self.start_batch,
+            "end_batch": self.end_batch,
+            "start_time": start_time,
+            "end_time": end_time,
+            "metrics": {},
+            "state": "STATE_ERRORED",
+        }
+
+        r = req.put(
+            self.master_url,
+            "/api/v1/training_metrics",
+            body={"training_metrics": training_metrics},
+        )
+        if not hasattr(r, "headers"):
+            raise Exception(r)
+
+        self.state = "STATE_ERRORED"
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):  # type: ignore
+        if self.state == "STATE_COMPLETED":
+            return
+
+        self.failed()
+
+
+class ValidationMetricManager:
```

Comment:
It's not code duplication. Checkpoints, steps, validations are independent data models that don't have a common ancestor.

Reply:
Yeah but there are 3 context managers all with the same interface, __enter__, __exit__, complete, failed. What I'm saying is that there duplication there in the interface is indicative of the fact they're not independent - they're all workloads, and these 3 interfaces all handle/act on them like they're workloads.

Reply:
You could have a `WorkloadManager` that knows how workloads are started and then either errored or completed and just let it know how to construct that particular workload type.

Reply:
We will be removing checkpointing/steps workloads. Also, the future workload won't be associated with just one checkpoint, validation, or training/step metric. So it doesn't make sense to have `BaseWorkloadManager`, which looks like the current data model.

Reply:
> We will be removing checkpointing/steps workloads.

I mean, we removed searcher emitted checkpoints but we won't be removing the concept of checkpoint workloads in general right?

> Also, the future workload won't be associated with just one checkpoint, validation, or training/step metric.

How could a workload be associated with more than one? I'm confused.

Reply:
> I mean, we removed searcher emitted checkpoints but we won't be removing the concept of checkpoint workloads in general right?

We will.

> How could a workload be associated with more than one? I'm confused.

Workloads are just an abstraction for things that the master tells the harness to do. In the ideal world, the harness registers checkpoints, validations, and training metrics itself rather than creates a workload, executes it, and return an artifact for each workload. We will still keep the workloads of validation, add a training workload, and remove the step and checkpoint workloads.